app/services/project_data_service.rb

```diff
@@ -52,8 +52,26 @@ def update_dependencies
   end
 
   def update_dependency(name, version)
+    update_required = dependency_update_required(name, version)
     @project.dependencies.find_or_initialize_by(name: name)
-            .update!(version: version)
+            .update!(version: version, update_required: update_required)
+  end
+
+  def dependency_update_required(name, version)
+    expectation = Horizon.config.stringify_keys["expected_version_#{name}"].split('.')
+    expected_major = expectation[0].to_i
+    expected_minor = expectation[1]&.to_i
+    return unless expected_major && version
+
+    current = version.delete('^0-9.').split('.')
+    current_major = current[0]&.to_i
+    current_minor = current[1]&.to_i
+    return unless current_major
+
+    return true if current_major < expected_major
+    return true if expected_minor && current_minor < expected_minor
+
+    false
```

Comment:
Comparing semantic version strings is non-trivial and we might be able to take advantage of prior art, such as in `Gem::Version` or [`semantic`](https://github.com/jlindsey/semantic). This would handle special cases like betas, and maybe even support matching against multiple constraints (like minimum _and_ maximum versions).

Reply:
I was able to get this halfway with the `semantic` gem-- unfortunately we use a variety of conventions in our project's version files, and dependencies frequently fail with errors like this: `12.x.x is not a valid SemVer Version (http://semver.org)`.

I updated for it to use `semantic` first to compare when `current` is a valid SemVer, and left my more manual parsing for the moment. If we enforce some naming conventions we could potentially remove this, but most of our versions fail without.